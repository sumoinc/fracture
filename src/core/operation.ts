import { paramCase } from "change-case";
import { Component } from "projen";
import { ValueOf } from "type-fest";
import { FractureService } from "./fracture-service";
import { Structure, StructureOptions } from "./structure";
import { StructureAttributeOptions } from "./structure-attribute";
import { DynamoGsi } from "../dynamodb";

/******************************************************************************
 * TYPES
 *****************************************************************************/

export const OperationType = {
  QUERY: "Query",
  MUTATION: "Mutation",
  SUBSCRIPTION: "Subscription",
} as const;

export const OperationSubType = {
  CREATE_ONE: "CreateOne",
  READ_ONE: "ReadOne",
  UPDATE_ONE: "UpdateOne",
  DELETE_ONE: "DeleteOne",
  IMPORT_ONE: "ImportOne",
  LIST: "LIST",
  CREATE_VERSION: "CreateVersion",
  READ_VERSION: "ReadVersion",
} as const;

export type OperationOptions = {
  /**
   * Name for this operation.
   */
  name: string;
  /**
   * Global secondary index used in this operation. This might also be the
   * primary PK and SK index.
   */
  dynamoGsi: DynamoGsi;
  /**
   * What (generic) type of operation is this?
   * Default: query
   */
  operationType?: ValueOf<typeof OperationType>;
  /**
   * What (more specific) specific type of operation is this?
   * Default: list
   */
  operationSubType?: ValueOf<typeof OperationSubType>;
  inputAttributeOptions?: Array<StructureAttributeOptions>;
  outputAttributeOptions?: Array<StructureAttributeOptions>;
};

export class Operation extends Component {
  /**
   * Name for this operation.
   */
  public readonly name: string;
  /**
   * Global secondary index used in this operation. This might also be the
   * primary PK and SK index.
   */
  public readonly dynamoGsi: DynamoGsi;
  /**
   * What (generic) type of operation is this?
   * Default: query
   */
  public readonly operationType: ValueOf<typeof OperationType>;
  /**
   * What (more specific) specific type of operation is this?
   * Default: list
   */
  public readonly operationSubType: ValueOf<typeof OperationSubType>;
  public readonly inputStructure: Structure;
  public readonly outputStructure: Structure;
  /**
   * All structures for this resource.
   */
  public structures: Structure[] = [];

  constructor(service: FractureService, options: OperationOptions) {
    super(service);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.name = paramCase(options.name);
    this.dynamoGsi = options.dynamoGsi;
    this.operationSubType = options.operationSubType || OperationSubType.LIST;
    this.operationType = options.operationType || OperationType.QUERY;

    /***************************************************************************
     * Initialize data structures
     **************************************************************************/

    this.inputStructure = this.addStructure({
      name: `${this.name}-input`,
      attributeOptions: options.inputAttributeOptions ?? [],
    });

    this.outputStructure = this.addStructure({
      name: `${this.name}-output`,
      attributeOptions: options.outputAttributeOptions ?? [],
    });

    return this;
  }

  public addStructure(options: StructureOptions) {
    const service = this.project as FractureService;
    const structure = new Structure(service, options);
    this.structures.push(structure);
    return structure;
  }

  /**
   * Operation name, based on the naming strategy
   */
  /*
  public get name() {
    const resourceName =
      this.operationSubType === OPERATION_SUB_TYPE.LIST
        ? this.resource.pluralName
        : this.resource.name;
    const prefix =
      this.service.namingStrategy.operations.prefixes[this.operationSubType];
    const suffix =
      this.service.namingStrategy.operations.suffixes[this.operationSubType];

    return [prefix, resourceName, suffix]
      .filter((part) => part.length > 0)
      .join("-");
  }
  */

  // public get operationType() {
  //   return this.options.operationType;
  // }

  // public get operationSubType() {
  //   return this.options.operationSubType;
  // }

  // public get isWrite(): boolean {
  //   return this.isCreate || this.isUpdate || this.isImport;
  // }

  // public get isCreate(): boolean {
  //   return this.operationSubType === OPERATION_SUB_TYPE.CREATE_ONE;
  // }

  // public get isRead(): boolean {
  //   return this.operationSubType === OPERATION_SUB_TYPE.READ_ONE;
  // }

  // public get isUpdate(): boolean {
  //   return this.operationSubType === OPERATION_SUB_TYPE.UPDATE_ONE;
  // }

  // public get isDelete(): boolean {
  //   return this.operationSubType === OPERATION_SUB_TYPE.DELETE_ONE;
  // }

  // public get isImport(): boolean {
  //   return this.operationSubType === OPERATION_SUB_TYPE.IMPORT_ONE;
  // }

  /***************************************************************************
   *
   * INPUT / OUTPUT STRUCTURES
   *
   * We lazy load these since all the attributes may not exist on the resource
   * when this component is initially created.
   *
   **************************************************************************/

  // public get inputStructure(): Structure {
  //   if (!this._inputStructure) {
  //     this._inputStructure = new Structure(this.resource, {
  //       type: STRUCTURE_TYPE.INPUT,
  //       operation: this,
  //     });
  //   }
  //   return this._inputStructure;
  // }

  // public get outputStructure(): Structure {
  //   if (!this._outputStructure) {
  //     this._outputStructure = new Structure(this.resource, {
  //       type: STRUCTURE_TYPE.OUTPUT,
  //       operation: this,
  //     });
  //   }
  //   return this._outputStructure;
  // }

  // public get resource(): Resource {
  //   return this.accessPattern.resource;
  // }

  /*
  public get service(): Service {
    return this.resource.service;
  }
  */

  /*****************************************************************************
   *
   *  TYPESCRIPT HELPERS
   *
   ****************************************************************************/

  /*
  public get tsResponseTypeName() {
    return this.operationSubType === OPERATION_SUB_TYPE.LIST
      ? this.service.tsLlistResponseTypeName
      : this.service.tsResponseTypeName;
  }
  */
}
