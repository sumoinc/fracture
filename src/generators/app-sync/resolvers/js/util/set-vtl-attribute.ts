import { ResourceAttributeGenerator } from "../../../../../services/resource-attribute";
import { StructureAttribute } from "../../../../../services/structure-attribute";
import { TypescriptStrategy } from "../../../../ts/strategy";

export const setVtlAttribute = (attribute: StructureAttribute) => {
  const strategy = TypescriptStrategy.of(attribute.project);
  const generator = attribute.generator;
  const name = strategy.formatAttributeName(attribute.name);
  const shortName = strategy.formatAttributeName(attribute.shortName);

  // non-generated
  if (generator === ResourceAttributeGenerator.NONE) {
    return `"${shortName}": $util.dynamodb.toDynamoDBJson($ctx.args.input.${name}),`;
    // generated values
  } else {
    switch (generator) {
      case ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP:
        return `"${shortName}": "$util.time.nowISO8601()"`;
      case ResourceAttributeGenerator.GUID:
        return `"${shortName}": "$util.autoId()"`;
      case ResourceAttributeGenerator.TYPE:
        return `"${shortName}": "unknown"`;

      default:
        throw new Error(`Unknown generator: ${generator}`);
    }
  }
};
