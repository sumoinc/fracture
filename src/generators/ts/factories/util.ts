import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
