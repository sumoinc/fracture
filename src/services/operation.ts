import { paramCase } from "change-case";
import { Component } from "projen";
import { TypeScriptProject } from "projen/lib/typescript";
import { ValueOf } from "type-fest";
import { DataService } from "./data-service";
import { Resource } from "./resource";
import {
  IdentifierType,
  ManagementType,
  ResourceAttribute,
  ResourceAttributeGenerator,
  ResourceAttributeType,
} from "./resource-attribute";
import { Structure, StructureOptions } from "./structure";
import {
  StructureAttribute,
  StructureAttributeOptions,
} from "./structure-attribute";
import { DynamoGsi, DynamoTable } from "../dynamodb";

/******************************************************************************
 * TYPES
 *****************************************************************************/

export const OperationType = {
  QUERY: "Query",
  MUTATION: "Mutation",
  SUBSCRIPTION: "Subscription",
} as const;

export const OperationSubType = {
  CREATE_ONE: "CreateOne",
  READ_ONE: "ReadOne",
  UPDATE_ONE: "UpdateOne",
  DELETE_ONE: "DeleteOne",
  IMPORT_ONE: "ImportOne",
  LIST: "LIST",
  CREATE_VERSION: "CreateVersion",
  READ_VERSION: "ReadVersion",
} as const;

export type OperationOptions = {
  /**
   *  Resource this structure belongs to.
   */
  readonly resource: Resource;

  /**
   * Name for this operation.
   */
  readonly name: string;

  /**
   * Global secondary index used in this operation. This might also be the
   * primary PK and SK index.
   *
   * @default - keyGsi (aka: no index)
   */
  readonly dynamoGsi?: DynamoGsi;

  /**
   * What (generic) type of operation is this?
   * Default: query
   */
  readonly operationType?: ValueOf<typeof OperationType>;

  /**
   * What (more specific) specific type of operation is this?
   * Default: list
   */
  readonly operationSubType?: ValueOf<typeof OperationSubType>;
  readonly inputAttributeOptions?: Array<StructureAttributeOptions>;
  //readonly outputAttributeOptions?: Array<StructureAttributeOptions>;
};

export class Operation extends Component {
  /**
   * Returns an attribute by name, or undefined if it doesn't exist
   */
  public static byName(
    resource: Resource,
    name: string
  ): Operation | undefined {
    const isDefined = (c: Component): c is Operation =>
      c instanceof Operation && c.name === name;
    return resource.operations.find(isDefined);
  }

  /**
   * Returns all structures for service
   */
  public static all(project: TypeScriptProject): Array<Operation> {
    const isDefined = (c: Component): c is Operation => c instanceof Operation;
    return project.components.filter(isDefined);
  }

  /**
   * Adds a create operation to the provided resource or returns existing
   * operation, if it exists
   */
  public static create(resource: Resource, name?: string): Operation {
    // does one exist al;ready?
    const isDefined = (c: Component): c is Operation =>
      c instanceof Operation &&
      c.operationSubType === OperationSubType.CREATE_ONE &&
      c.name === name;
    const foundOperation = resource.operations.find(isDefined);
    if (foundOperation) {
      return foundOperation;
    }

    // create one, then return it
    const operationName = name || `create-${resource.name}`;
    const o = new Operation(resource.service, {
      name: operationName,
      resource,
      operationType: OperationType.MUTATION,
      operationSubType: OperationSubType.CREATE_ONE,
      dynamoGsi: DynamoTable.of(resource.project).keyGsi,
    });

    resource.attributes.forEach((attribute) => {
      // all identifiers are required
      if (attribute.identifier === IdentifierType.PRIMARY) {
        o.addInputAttribute(attribute);
      }
      // add all user namaged attributes
      if (
        attribute.management === ManagementType.USER_MANAGED &&
        attribute.type !== ResourceAttributeType.ARRAY &&
        attribute.type !== ResourceAttributeType.MAP
      ) {
        o.addInputAttribute(attribute);
      }
    });

    return o;
  }

  /**
   * Adds a read operation to the provided resource or returns existing
   * operation, if it exists
   */
  public static read(resource: Resource, name?: string): Operation {
    // does one exist al;ready?
    const isDefined = (c: Component): c is Operation =>
      c instanceof Operation &&
      c.operationSubType === OperationSubType.READ_ONE &&
      c.name === name;
    const foundOperation = resource.operations.find(isDefined);
    if (foundOperation) {
      return foundOperation;
    }

    // create one, then return it
    const operationName = name || `get-${resource.name}`;
    const o = new Operation(resource.service, {
      name: operationName,
      resource,
      operationType: OperationType.QUERY,
      operationSubType: OperationSubType.READ_ONE,
      dynamoGsi: DynamoTable.of(resource.project).keyGsi,
    });

    resource.attributes.forEach((attribute) => {
      // all identifiers are required
      if (attribute.identifier === IdentifierType.PRIMARY) {
        o.addInputAttribute(attribute);
      }
    });

    return o;
  }

  /**
   * Adds a update operation to the provided resource or returns existing
   * operation, if it exists
   */
  public static update(resource: Resource, name?: string): Operation {
    // does one exist al;ready?
    const isDefined = (c: Component): c is Operation =>
      c instanceof Operation &&
      c.operationSubType === OperationSubType.UPDATE_ONE &&
      c.name === name;
    const foundOperation = resource.operations.find(isDefined);
    if (foundOperation) {
      return foundOperation;
    }

    // create one, then return it
    const operationName = name || `get-${resource.name}`;
    const o = new Operation(resource.service, {
      name: operationName,
      resource,
      operationType: OperationType.QUERY,
      operationSubType: OperationSubType.READ_ONE,
      dynamoGsi: DynamoTable.of(resource.project).keyGsi,
    });

    resource.attributes.forEach((attribute) => {
      // all identifiers are required
      if (attribute.identifier === IdentifierType.PRIMARY) {
        o.addInputAttribute(attribute);
      }
      // add all user namaged attributes
      if (
        attribute.management === ManagementType.USER_MANAGED &&
        attribute.type !== ResourceAttributeType.ARRAY &&
        attribute.type !== ResourceAttributeType.MAP
      ) {
        o.addInputAttribute(attribute, {
          required: false,
        });
      }
    });

    return o;
  }

  /**
   * Adds a delete operation to the provided resource or returns existing
   * operation, if it exists
   */
  public static delete(resource: Resource, name?: string): Operation {
    // does one exist al;ready?
    const isDefined = (c: Component): c is Operation =>
      c instanceof Operation &&
      c.operationSubType === OperationSubType.DELETE_ONE &&
      c.name === name;
    const foundOperation = resource.operations.find(isDefined);
    if (foundOperation) {
      return foundOperation;
    }

    // create one, then return it
    const operationName = name || `get-${resource.name}`;
    const o = new Operation(resource.service, {
      name: operationName,
      resource,
      operationType: OperationType.QUERY,
      operationSubType: OperationSubType.READ_ONE,
      dynamoGsi: DynamoTable.of(resource.project).keyGsi,
    });

    resource.attributes.forEach((attribute) => {
      // all identifiers are required
      if (attribute.identifier === IdentifierType.PRIMARY) {
        o.addInputAttribute(attribute);
      }
    });

    return o;
  }

  /**
   *  Resource this structure belongs to.
   */
  public readonly resource: Resource;

  /**
   * Name for this operation.
   */
  public readonly name: string;

  /**
   * Global secondary index used in this operation. This might also be the
   * primary PK and SK index.
   *
   *  @default - keyGsi (aka: no index)
   */
  public readonly dynamoGsi: DynamoGsi;

  /**
   * What (generic) type of operation is this?
   * Default: query
   */
  public readonly operationType: ValueOf<typeof OperationType>;

  /**
   * What (more specific) specific type of operation is this?
   * Default: list
   */
  public readonly operationSubType: ValueOf<typeof OperationSubType>;
  public readonly inputStructure: Structure;
  public readonly outputAttribute: StructureAttribute;

  /**
   * All structures for this resource.
   */
  public structures: Array<Structure> = [];

  constructor(public readonly project: DataService, options: OperationOptions) {
    super(project);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.resource = options.resource;
    this.name = paramCase(options.name);
    (this.dynamoGsi = options.dynamoGsi || DynamoTable.of(this.service).keyGsi),
      (this.operationSubType =
        options.operationSubType || OperationSubType.LIST);
    this.operationType = options.operationType || OperationType.QUERY;

    /***************************************************************************
     * Initialize data structures
     **************************************************************************/

    this.inputStructure = this.addStructure({
      name: `${this.name}-input`,
      attributeOptions: options.inputAttributeOptions ?? [],
    });

    this.outputAttribute = new StructureAttribute(this.project, {
      name: this.resource.name,
      shortName: this.resource.shortName,
      type: this.resource.name,
    });

    /*
    this.outputStructure = this.addStructure({
      name: `${this.name}-output`,
      attributeOptions: options.outputAttributeOptions ?? [],
    });

    */

    /***************************************************************************
     * Add to Resource
     **************************************************************************/

    if (Operation.byName(this.resource, this.name)) {
      throw new Error(
        `Resource "${this.resource.name}" already has an operation named "${this.name}"`
      );
    }

    this.resource.operations.push(this);

    return this;
  }

  public addStructure(options: StructureOptions) {
    const structure = new Structure(this.project, options);
    this.structures.push(structure);
    return structure;
  }

  public addInputAttribute(
    attribute: ResourceAttribute,
    options: Partial<StructureAttributeOptions> = {}
  ) {
    // apply any generators?
    const generator =
      this.operationSubType === OperationSubType.CREATE_ONE
        ? attribute.createGenerator
        : this.operationSubType === OperationSubType.READ_ONE
        ? attribute.readGenerator
        : this.operationSubType === OperationSubType.UPDATE_ONE
        ? attribute.updateGenerator
        : this.operationSubType === OperationSubType.DELETE_ONE
        ? attribute.deleteGenerator
        : ResourceAttributeGenerator.NONE;

    return this.inputStructure.addAttribute({
      name: attribute.name,
      shortName: attribute.shortName,
      type: attribute.type,
      typeParameter: attribute.typeParameter,
      comments: attribute.comments,
      required: attribute.required,
      generator,
      ...options,
    });
  }

  public get service() {
    return this.project;
  }

  /***************************************************************************
   * Configuration export for this operation
   **************************************************************************/

  public config(): Record<string, any> {
    return {
      name: this.name,
      dynamoGsi: this.dynamoGsi,
      operationType: this.operationType,
      operationSubType: this.operationSubType,
      inputStructure: this.inputStructure.config(),
      outputAttribute: this.outputAttribute.config(),
    };
  }
}
