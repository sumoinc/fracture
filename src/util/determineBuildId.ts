import { execSync } from "child_process";

/**
 * Returns the current git branch name, without folder.
 *
 * ie: feature/1234 returns 1234
 *
 * @returns {string}
 */
export const determineBuildId = () => {
  const branchName = execSync("git rev-parse --abbrev-ref HEAD")
    .toString("utf8")
    .replace(/[\n\r\s]+$/, "");
  const branchPath = branchName.split("/");
  return branchPath[branchPath.length - 1];
};
