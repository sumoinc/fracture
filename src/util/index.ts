export * from "./determineBuildId";
export * from "./test-util";
