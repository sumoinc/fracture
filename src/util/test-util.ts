import { Project } from "projen";
import { synthSnapshot } from "projen/lib/util/synth";

export const TEST_ACCOUNT_ONE = "000000000000";
export const TEST_ORG_ONE = "org-123456";
export const TEST_REGION_ONE = "us-east-1";

/**
 *
 * Helper to generate and return one file from fracture.
 * Useful when trying to test the final output of generated files.
 *
 * @param fracture
 * @param filepath
 * @returns collection of files starting with supplied filepath
 */
export const synthFiles = (project: Project, filepath: string = ""): any => {
  const snapshot = synthSnapshot(project);
  // console.log(Object.keys(snapshot));
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(filepath))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {} as { [key: string]: any });
  return filtered;
};

export const synthFile = (project: Project, filepath: string): string => {
  const files = synthFiles(project, filepath);

  if (files.length > 1) {
    throw new Error(`more than one file matched filepath "${filepath}"`);
  }

  if (files.length === 0) {
    throw new Error(`No files matched filepath "${filepath}"`);
  }

  return files[filepath];
};

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testProject = () => {
//   return new Project({
//     name: "my-project",
//   });
// };

// export const testNodeProject = (options: Partial<NodeProjectOptions> = {}) => {
//   return new NodeProject({
//     name: "my-project",
//     defaultReleaseBranch: "main",
//     packageManager: NodePackageManager.PNPM,
//     pnpmVersion: "8",
//     ...options,
//   });
// };

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testFractureProject = (
//   options: Partial<FractureProjectOptions> = {}
// ) => {
//   const fp = new FractureProject({
//     name: "my-project",
//     ...options,
//   });

//   /*
//   fp.buildEnvironment = new AwsEnvironment(fp, {
//     name: "test",
//     account: TEST_ACCOUNT_ONE,
//     region: TEST_REGION_ONE,
//     primaryHostedZoneId: "foo",
//     primaryDomainName: "bar",
//   });
//   */

//   return fp;
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testApp = () => {
//   return new App({
//     parent: testFractureProject(),
//     name: "my-app",
//   });
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testDataService = () => {
//   return new DataService(testApp(), {
//     name: "my-service",
//   });
// };

// /**
//  * Build a simple resource to test with
//  */
// export const testResource = () => {
//   const service = testDataService();
//   return new Resource(service.app, {
//     service,
//     name: "my-resource",
//   });
// };

// /**
//  * A more complicated resource configuration
//  */
// export const complexService = () => {
//   const service = testDataService();

//   const person = new Resource(service.app, { service, name: "person" });

//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     comments: ["A club"],
//     attributeOptions: [{ name: "name" }],
//   });

//   person.addAttribute({
//     name: "first-name",
//     shortName: "fn",
//   });
//   person.addAttribute({
//     name: "last-name",
//     shortName: "ln",
//   });
//   person.addAttribute({
//     name: "height",
//     shortName: "ht",
//     type: ResourceAttributeType.INT,
//   });
//   person.addAttribute({
//     name: "is-frog",
//     shortName: "if",
//     type: ResourceAttributeType.BOOLEAN,
//     comments: ["Is this person a frog?"],
//   });
//   person.addAttribute({
//     name: "fav-colors",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: ResourceAttributeType.STRING,
//     comments: ["List opf favorite colors"],
//   });

//   // one specific club
//   person.addAttribute({
//     name: "one-club",
//     type: club,
//   });
//   // an array of clubs
//   person.addAttribute({
//     name: "array-of-clubs",
//     shortName: "ad",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });

//   // an array of persons
//   club.addAttribute({
//     name: "array-of-persons",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });

//   // operation
//   Operation.create(person);
//   Operation.read(person);
//   Operation.update(person);
//   Operation.delete(person);

//   return service;
// };

// /**
//  * Build a simple resource attribute to test with
//  */
// export const testResourceAttribute = () => {
//   const service = testDataService();
//   const resource = testResource();
//   return new ResourceAttribute(service.app, {
//     name: "my-attr",
//     resource,
//   });
// };

// /**
//  * Build a simple structure to test with
//  */
// export const testStructure = () => {
//   const service = testDataService();
//   return new Structure(service.app, { name: "my-structure" });
// };

// /**
//  * Build a simple operation to test with
//  */
// export const testOperation = () => {
//   const resource = testResource();
//   return new Operation(resource.service.app, {
//     name: "my-operation",
//     resource,
//   });
// };

// /**
//  * Build a nuxt site to test with
//  */
// export const testNuxtJsSite = () => {
//   return new NuxtJsSite({
//     parent: testFractureProject(),
//     name: "my-site",
//   });
// };
// export const testServiceTwoRelatedResources = () => {
//   const service = testDataService();
//   const person = new Resource(service.app, {
//     service,
//     name: "person",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   club.addAttribute({
//     name: "people",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });
//   person.addAttribute({
//     name: "clubs",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });
//   return { service, person, club };
// };
