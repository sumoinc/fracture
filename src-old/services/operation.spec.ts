import { Operation } from "./operation";
import { testResource } from "../util/test-util";

describe("success conditions", () => {
  test("Smoke test without helper", () => {
    const resource = testResource();
    const operation = new Operation(resource.service, {
      name: "save-user",
      resource,
    });

    // exists
    expect(operation).toBeTruthy();
    // also contained in array
    expect(
      resource.operations.findIndex((o) => {
        return o.name === o.name;
      })
    ).toBeGreaterThanOrEqual(0);
  });

  test("Smoke test with helper", () => {
    const resource = testResource();
    const operation = resource.addOperation({
      name: "save-user",
    });

    // exists
    expect(operation).toBeTruthy();
    // also contained in array
    expect(
      resource.operations.findIndex((o) => {
        return o.name === o.name;
      })
    ).toBeGreaterThanOrEqual(0);
  });
});

describe("crud operations", () => {
  test("Create Operation", () => {
    const operation = Operation.create(testOperationResource());

    // proper input
    expect(operation?.inputStructure).toBeTruthy();
    expect(
      JSON.stringify(operation?.inputStructure.config(), null, 2)
    ).toMatchSnapshot();

    // proper output
    expect(operation?.outputAttribute).toBeTruthy();
    expect(
      JSON.stringify(operation?.outputAttribute.config(), null, 2)
    ).toMatchSnapshot();

    //console.log(operation?.inputStructure.config());
    //console.log(operation?.outputAttribute.config());
  });

  test("Read Operation", () => {
    const operation = Operation.read(testOperationResource());

    // proper input
    expect(operation?.inputStructure).toBeTruthy();
    expect(
      JSON.stringify(operation?.inputStructure.config(), null, 2)
    ).toMatchSnapshot();

    // proper output
    expect(operation?.outputAttribute).toBeTruthy();
    expect(
      JSON.stringify(operation?.outputAttribute.config(), null, 2)
    ).toMatchSnapshot();

    //console.log(operation?.inputStructure.config());
    //console.log(operation?.outputAttribute.config());
  });

  test("Update Operation", () => {
    const operation = Operation.update(testOperationResource());

    // proper input
    expect(operation?.inputStructure).toBeTruthy();
    expect(
      JSON.stringify(operation?.inputStructure.config(), null, 2)
    ).toMatchSnapshot();

    // proper output
    expect(operation?.outputAttribute).toBeTruthy();
    expect(
      JSON.stringify(operation?.outputAttribute.config(), null, 2)
    ).toMatchSnapshot();

    // console.log(operation?.inputStructure.config());
    // console.log(operation?.outputAttribute.config());
  });

  test("Delete Operation", () => {
    const operation = Operation.delete(testOperationResource());

    // proper input
    expect(operation?.inputStructure).toBeTruthy();
    expect(
      JSON.stringify(operation?.inputStructure.config(), null, 2)
    ).toMatchSnapshot();

    // proper output
    expect(operation?.outputAttribute).toBeTruthy();
    expect(
      JSON.stringify(operation?.outputAttribute.config(), null, 2)
    ).toMatchSnapshot();

    // console.log(operation?.inputStructure.config());
    // console.log(operation?.outputAttribute.config());
  });
});

// semi complext resource for testing CRUD operations
const testOperationResource = () => {
  const resource = testResource();
  resource.addAttribute({
    name: "required-name",
    required: true,
  });
  resource.addAttribute({
    name: "optional-name",
    required: false,
  });
  return resource;
};
