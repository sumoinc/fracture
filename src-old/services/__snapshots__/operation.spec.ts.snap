// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crud operations Create Operation 1`] = `
"{
  \\"name\\": \\"create-my-resource-input\\",
  \\"comments\\": [],
  \\"persistant\\": true,
  \\"attributes\\": [
    {
      \\"name\\": \\"required-name\\",
      \\"shortName\\": \\"required-name\\",
      \\"type\\": \\"String\\",
      \\"management\\": \\"USER_MANAGED\\",
      \\"visibility\\": \\"USER_VISIBLE\\",
      \\"comments\\": [],
      \\"required\\": true,
      \\"generator\\": \\"None\\"
    },
    {
      \\"name\\": \\"optional-name\\",
      \\"shortName\\": \\"optional-name\\",
      \\"type\\": \\"String\\",
      \\"management\\": \\"USER_MANAGED\\",
      \\"visibility\\": \\"USER_VISIBLE\\",
      \\"comments\\": [],
      \\"required\\": false,
      \\"generator\\": \\"None\\"
    }
  ]
}"
`;

exports[`crud operations Create Operation 2`] = `
"{
  \\"name\\": \\"my-resource\\",
  \\"shortName\\": \\"my-resource\\",
  \\"type\\": \\"my-resource\\",
  \\"management\\": \\"SYSTEM_MANAGED\\",
  \\"visibility\\": \\"USER_VISIBLE\\",
  \\"comments\\": [],
  \\"required\\": true,
  \\"generator\\": \\"None\\"
}"
`;

exports[`crud operations Delete Operation 1`] = `
"{
  \\"name\\": \\"delete-my-resource-input\\",
  \\"comments\\": [],
  \\"persistant\\": true,
  \\"attributes\\": [
    {
      \\"name\\": \\"id\\",
      \\"shortName\\": \\"id\\",
      \\"type\\": \\"String\\",
      \\"management\\": \\"SYSTEM_MANAGED\\",
      \\"visibility\\": \\"USER_VISIBLE\\",
      \\"comments\\": [
        \\"Identifier for this record.\\"
      ],
      \\"required\\": true,
      \\"generator\\": \\"None\\"
    }
  ]
}"
`;

exports[`crud operations Delete Operation 2`] = `
"{
  \\"name\\": \\"my-resource\\",
  \\"shortName\\": \\"my-resource\\",
  \\"type\\": \\"my-resource\\",
  \\"management\\": \\"SYSTEM_MANAGED\\",
  \\"visibility\\": \\"USER_VISIBLE\\",
  \\"comments\\": [],
  \\"required\\": true,
  \\"generator\\": \\"None\\"
}"
`;

exports[`crud operations Read Operation 1`] = `
"{
  \\"name\\": \\"get-my-resource-input\\",
  \\"comments\\": [],
  \\"persistant\\": true,
  \\"attributes\\": [
    {
      \\"name\\": \\"id\\",
      \\"shortName\\": \\"id\\",
      \\"type\\": \\"String\\",
      \\"management\\": \\"SYSTEM_MANAGED\\",
      \\"visibility\\": \\"USER_VISIBLE\\",
      \\"comments\\": [
        \\"Identifier for this record.\\"
      ],
      \\"required\\": true,
      \\"generator\\": \\"None\\"
    }
  ]
}"
`;

exports[`crud operations Read Operation 2`] = `
"{
  \\"name\\": \\"my-resource\\",
  \\"shortName\\": \\"my-resource\\",
  \\"type\\": \\"my-resource\\",
  \\"management\\": \\"SYSTEM_MANAGED\\",
  \\"visibility\\": \\"USER_VISIBLE\\",
  \\"comments\\": [],
  \\"required\\": true,
  \\"generator\\": \\"None\\"
}"
`;

exports[`crud operations Update Operation 1`] = `
"{
  \\"name\\": \\"update-my-resource-input\\",
  \\"comments\\": [],
  \\"persistant\\": true,
  \\"attributes\\": [
    {
      \\"name\\": \\"id\\",
      \\"shortName\\": \\"id\\",
      \\"type\\": \\"String\\",
      \\"management\\": \\"SYSTEM_MANAGED\\",
      \\"visibility\\": \\"USER_VISIBLE\\",
      \\"comments\\": [
        \\"Identifier for this record.\\"
      ],
      \\"required\\": true,
      \\"generator\\": \\"None\\"
    },
    {
      \\"name\\": \\"required-name\\",
      \\"shortName\\": \\"required-name\\",
      \\"type\\": \\"String\\",
      \\"management\\": \\"USER_MANAGED\\",
      \\"visibility\\": \\"USER_VISIBLE\\",
      \\"comments\\": [],
      \\"required\\": false,
      \\"generator\\": \\"None\\"
    },
    {
      \\"name\\": \\"optional-name\\",
      \\"shortName\\": \\"optional-name\\",
      \\"type\\": \\"String\\",
      \\"management\\": \\"USER_MANAGED\\",
      \\"visibility\\": \\"USER_VISIBLE\\",
      \\"comments\\": [],
      \\"required\\": false,
      \\"generator\\": \\"None\\"
    }
  ]
}"
`;

exports[`crud operations Update Operation 2`] = `
"{
  \\"name\\": \\"my-resource\\",
  \\"shortName\\": \\"my-resource\\",
  \\"type\\": \\"my-resource\\",
  \\"management\\": \\"SYSTEM_MANAGED\\",
  \\"visibility\\": \\"USER_VISIBLE\\",
  \\"comments\\": [],
  \\"required\\": true,
  \\"generator\\": \\"None\\"
}"
`;
